sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBSuite.scala

```diff
@@ -2240,6 +2240,166 @@ class RocksDBSuite extends AlsoTestWithChangelogCheckpointingEnabled with Shared
     }
   }
 
+  test("Rocks DB task completion listener does not double unlock acquireThread") {
+    // Create a custom ExecutionContext with 3 threads
+    implicit val ec: ExecutionContext = ExecutionContext.fromExecutor(
+      ThreadUtils.newDaemonFixedThreadPool(5, "pool-thread-executor"))
+    val timeout = 5.seconds
+
+    val stateLock = new Object()
+    var state = 0
+    // Timeline of this test:
+    // STATE | MAIN             | THREAD 1         | THREAD 2         | THREAD 3
+    // ------| ---------------- | ---------------- | ---------------- | ----------------
+    // 0.    | wait for s2      | Load, commit     |                  |
+    //       |                  | wait for s1      | load, signal s1  |
+    // 1.    |                  | task complete    | wait for s3      |
+    //       |                  | signal s2, END   |                  |
+    // 2.    | start thread 3   |                  |                  | load, commit
+    //       | wait for timeout |                  |                  | timeout, END
+    //       | signal s3        |                  |                  |
+    // 3.    | wait for s4      |                  | commit           |
+    //       |                  |                  | signal s4, END   |
+    // 4.    | close db, END    |                  |                  |
+
+    withTempDir { dir =>
+      val remoteDir = dir.getCanonicalPath
+      val conf = dbConf.copy(lockAcquireTimeoutMs = timeout.toMillis.toInt)
+      val db = new RocksDB(
+        remoteDir,
+        conf = conf,
+        localRootDir = Utils.createTempDir(),
+        hadoopConf = new Configuration(),
+        loggingId = s"[Thread-${Thread.currentThread.getId}]",
+        useColumnFamilies = false
+      )
+      try {
+        Future { // THREAD 1
+          // Set thread 1's task context so that it is not a clone
+          // of withDB's taskContext, which will close the db when the
+          // task is marked as complete
+          val taskContext = TaskContext.empty()
+          TaskContext.setTaskContext(taskContext)
+
+          // Simulate a task that loads and commits, db should be unlocked after
+          db.load(0)
+          db.put("a", "1")
+          db.commit()
+
+          Future { // THREAD 2
+            // Set thread 2's task context so that it is not a clone of thread 1's
+            // so it won't be marked as complete
+            val taskContext = TaskContext.empty()
+            TaskContext.setTaskContext(taskContext)
+
+            // Load the db and signal that we have entered state 1
+            stateLock.synchronized {
+              db.load(1)
+              state = 1
+              stateLock.notifyAll()
+            }
+
+            stateLock.synchronized {
+              // Wait until we have entered state 3, commit and signal
+              // that we have entered state 4
+              while (state != 3) {
+                stateLock.wait()
+              }
```

Comment:
I don't see assertions in this test, to confirm our assumptions. Do we need to assert here that the current acquiredThreadInfo is actually for this thread? To confirm that it wasn't released. We should do that in other places needed to.

Because for example in this place, if the lock has been accidentally released, commit() will still work since we don't acquire lock in commit and just release the lock acquired in load.